### PR TITLE
planner: BatchPointGet support Update and Delete on partitioned tables.

### PR DESCRIPTION
**Pull request by coocood**
### What problem does this PR solve? 
make BatchPointGet support Update and Delete on partitioned tables.

### What is changed and how it works?
Since the normal path planner first transforms a partitioned Datasource to Union, we can not support BatchPointGet for partitioned tables.
So we build the BatchPointGet plan on the fast path to support Update/Delete.

### Check List 

Tests 

 - Unit test

Release note

 - BatchPointGet supports Update and Delete on partitioned tables.

### CODE REVIEW

**Comment by coocood** [comment]
@tiancaiamao @lzmhhh123 PTAL

**Comment by siddontang** [comment]
can we use partition for TPCC now?

**Comment by coocood** [comment]
@siddontang yes, after this PR is merged.

**Comment by coocood** [comment]
Should be merged after https://github.com/pingcap/tidb/issues/15361 has been fixed.

**Comment by coocood** [comment]
@tiancaiamao @lzmhhh123 PTAL

File: planner/core/point_get_plan.go [hunk 1]
```diff
@@ -330,11 +330,13 @@ func TryFastPlan(ctx sessionctx.Context, node ast.Node) Plan {
 		// Try to convert the `SELECT a, b, c FROM t WHERE (a, b, c) in ((1, 2, 4), (1, 3, 5))` to
 		// `PhysicalUnionAll` which children are `PointGet` if exists an unique key (a, b, c) in table `t`
 		if fp := tryWhereIn2BatchPointGet(ctx, x); fp != nil {
+			if checkFastPlanPrivilege(ctx, fp.dbName, fp.TblInfo.Name.L, mysql.SelectPriv) != nil {
+				return nil
```
Comment (tiancaiamao) [inline planner/core/point_get_plan.go]: if err != nil return nil ???

File: planner/core/point_get_plan.go [hunk 2]
```diff
@@ -330,11 +330,13 @@ func TryFastPlan(ctx sessionctx.Context, node ast.Node) Plan {
...
+			}
 			fp.Lock, fp.LockWaitTime = getLockWaitTime(ctx, x.LockTp)
 			return fp
 		}
-		fp := tryPointGetPlan(ctx, x)
-		if fp != nil {
+		if fp := tryPointGetPlan(ctx, x); fp != nil {
 			if checkFastPlanPrivilege(ctx, fp.dbName, fp.TblInfo.Name.L, mysql.SelectPriv) != nil {
 				return nil
```
Comment (tiancaiamao) [inline planner/core/point_get_plan.go]: ditto

File: planner/core/point_get_plan.go [hunk 3]
```diff
@@ -945,58 +940,69 @@ func tryUpdatePointPlan(ctx sessionctx.Context, updateStmt *ast.UpdateStmt) Plan
...
+		return buildPointUpdatePlan(ctx, batchPointGet, batchPointGet.dbName, batchPointGet.TblInfo, updateStmt)
 	}
-	if ctx.GetSessionVars().TxnCtx.IsPessimistic {
-		fastSelect.Lock = true
+	return nil
+}
+
+func buildPointUpdatePlan(ctx sessionctx.Context, pointPlan PhysicalPlan, dbName string, tbl *model.TableInfo, updateStmt *ast.UpdateStmt) Plan {
+	if checkFastPlanPrivilege(ctx, dbName, tbl.Name.L, mysql.SelectPriv, mysql.UpdatePriv) != nil {
+		return nil
```
Comment (tiancaiamao) [inline planner/core/point_get_plan.go]: ditto

File: planner/core/point_get_plan.go [hunk 4]
```diff
@@ -1026,29 +1032,38 @@ func tryDeletePointPlan(ctx sessionctx.Context, delStmt *ast.DeleteStmt) Plan {
...
+		return buildPointDeletePlan(ctx, batchPointGet, batchPointGet.dbName, batchPointGet.TblInfo)
 	}
-	if ctx.GetSessionVars().TxnCtx.IsPessimistic {
-		fastSelect.Lock = true
+	return nil
+}
+
+func buildPointDeletePlan(ctx sessionctx.Context, pointPlan PhysicalPlan, dbName string, tbl *model.TableInfo) Plan {
+	if checkFastPlanPrivilege(ctx, dbName, tbl.Name.L, mysql.SelectPriv, mysql.DeletePriv) != nil {
+		return nil
```
Comment (tiancaiamao) [inline planner/core/point_get_plan.go]: ditto

File: planner/core/point_get_plan.go (hunk 1 above)
Comment (coocood) [reply planner/core/point_get_plan.go]: We can fall back to the normal plan, then return the error then.

**Comment by tiancaiamao** [comment]
LGTM

File: planner/core/point_get_plan.go (hunk 1 above)
Comment (tiancaiamao) [reply planner/core/point_get_plan.go]: Get it!

**Pull request merged**